### PR TITLE
✨ Feature ➾ Allow docker images to be overwritten via environment variable (#1425)

### LINKED ISSUE #1425: 🚧 Dev ➾ Provide a means of overriding all docker images
### 🔗 Epic|Feature Link

_No response_

### 🚥 Definition of Done

- [x] 🎯 Planning and Design
- [ ] 🏎️ Dev Implementation Complete
- [ ] ♾️ PR Merged
- [ ] 🏆 Product Acceptance

### ✅ To Do

- [ ] #1508
- [ ] Determine whether each plugin that uses a docker image should provide a "show-image-name" task.

### ⚖️ Acceptance Criteria

✨ Functional: 
- [ ] Add a function to the Taqueria SDK like, `getDockerimage(imageName: string, envVarName: string)`
- [ ] Adjust ligo plugin to use this method
- [ ] Adjust archetype plugin to use this method
- [ ] While we're editing relevant code, we may as well adjust any image names to use the latest version tag of the docker image.

### 🔭 Scope

Associated ticket: https://github.com/ecadlabs/taqueria/issues/1425

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by mweichert**
# 🌮 Taqueria PR

## 🪁 Description

There are four plugins currently which use docker images to facilitate their functionality:
- ligo plugin
- archetype plugin
- flextesa plugin
- tezos-client plugin

## 🪂 Pre-Merge Checklist (Definition of Done)

🚦 Required to merge:

- [x] ⛱️ I have completed this PR template in full and updated the title appropriately
- [x] ⛵ My code builds cleanly, and I have manually tested the changes
- [x] 🏄‍♂️ Another team member has built this branch and done manual testing on the change
- [x] 🏖️ New and existing unit tests pass locally and in CI
- [x] 🔱 The test plan has been implemented and verified by an SDET
- [x] 🦀 Automated tests have been written and added to this PR
- [x] 🐬 I have commented my code, particularly in hard-to-understand areas
- [ ] 🤿 Corresponding changes have been made to all documentation
- [x] 🐚 Required changes to the VScE have been made
- [x] 🪸 Required updates to scaffolds have been made
- [ ] 🚢 The release checklist has been completed

## 🛩️ Summary of Changes

Users were looking for the ability to override the docker image being used. This is a common request in CICD, and thus we opted to allow overrides using environment variables. This is not mutually exclusive from adding an option later to allow specifying version #s in the config file, such as the ligo version.

## 🎢 Test Plan

- I introduced a new function in the SDK called `getDockerImage`, and added a unit test for it in taqueria-sdk.spec.ts
- For each plugin, I added a test case to assure that each plugin could have a custom image specified using a plugin. To do this, I added a hidden task for each plugin called `get-image` that our tests can use to verify what image is being used. 

## 🛸 Type of Change

- [ ] 🧽 Chore ➾
- [ ] 🛠️ Fix ➾
- [x] ✨ Feature ➾
- [ ] 👷 Refactor ➾
- [ ] 🧪 Pre-Release ➾
- [ ] 🚀 Release ➾

### CODE REVIEW

**Comment by AlirezaHaghshenas** [comment]
One general question?
Why do we enable this through environment variables and not by optional settings in `config.json`?

**Review by jchenche — COMMENTED** [review]
Some e2e tests are failing. Can we address those first?

File: taqueria-plugin-ligo/common.ts [hunk 1]
```diff
@@ -14,7 +15,7 @@ export type IntersectionOpts = CompileOpts & TestOpts;
 type UnionOpts = CompileOpts | TestOpts;
 
 // Points to the latest stable version. Needs to update this as part of our release process
-export const LIGO_DOCKER_IMAGE = 'ligolang/ligo:0.51.0';
+export const LIGO_DOCKER_IMAGE = getDockerImage('ligolang/ligo:0.51.0', 'TAQ_LIGO_IMAGE');
```
Comment (jchenche) [inline taqueria-plugin-ligo/common.ts:18]: Doesn't sound like good practice to have a function return a value that may be different with each invocation and assign it to a global constant.

**Comment by mweichert** [comment]
> One general question? Why do we enable this through environment variables and not by optional settings in `config.json`?

@AlirezaHaghshenas somehow I missed your comment here. It was thought that this would be easiest for CI/CD setups. However, I don't think this is mutually exclusive from also offering the ability override by config. I think that's a great idea too. But this was just meant to resolve a particular use-case.

File: taqueria-plugin-ligo/common.ts (hunk 1 above)
Comment (mweichert) [reply taqueria-plugin-ligo/common.ts:18]: I see this as effectively the same as a constant value - it just can be temporary overwritten for a particular invocation. The value doesn't change over the lifetime of the execution, so I thought this was fine. Did you have something else in mind? I'm open to other approaches.

Comment (jchenche) [reply taqueria-plugin-ligo/common.ts:22]: I think it'd be better to use `getDockerImage()` when we are using the image. And then the function would depend on LIGO_DOCKER_IMAGE as a parameter.

e.g. `getDockerImage(LIGO_DOCKER_IMAGE, 'TAQ_LIGO_IMAGE')`

File: taqueria-plugin-ligo/compile.ts [hunk 2]
```diff
@@ -93,7 +93,8 @@ const compileContract = (parsedArgs: Opts, sourceFile: string): Promise<TableRow
 		})
 		.catch(err => {
 			sendErr(`\n=== For ${sourceFile} ===`);
-			sendErr(err.message.replace(/Command failed.+?\n/, ''));
+			sendErr(err.toString());
```
Comment (jchenche) [inline taqueria-plugin-ligo/compile.ts:96]: Seems like we are replicating the error message here. Can we revert this?

**Review by jchenche — CHANGES_REQUESTED** [review]
This PR is outdated. Can you merge main in and resolve all conflicts?

File: taqueria-plugin-ligo/common.ts [hunk 3]
```diff
@@ -18,7 +19,7 @@ export type IntersectionOpts = LigoOpts & CompileOpts & TestOpts;
 type UnionOpts = LigoOpts | CompileOpts | TestOpts;
 
 // Points to the latest stable version. Needs to update this as part of our release process
-export const LIGO_DOCKER_IMAGE = 'ligolang/ligo:0.55.0';
+export const LIGO_DOCKER_IMAGE = getDockerImage('ligolang/ligo:0.55.0', 'TAQ_LIGO_IMAGE');
```
Comment (jchenche) [inline taqueria-plugin-ligo/common.ts:22]: It is easier to reason and reduces cognitive overhead to only assign literals to global constants

**Review by jchenche — CHANGES_REQUESTED** [review]

**Review by AlirezaHaghshenas — APPROVED** [review]
LGTM.

I would personally be happier if we had not created the get-image tasks.

Is there an end-user use-case for those?

**Comment by mweichert** [comment]
> LGTM.
> 
> I would personally be happier if we had not created the get-image tasks.
> 
> Is there an end-user use-case for those?

It's just to verify what docker image is used by a particular plugin, just in case you need to double check. I believe I made these tasks hidden from the UI.

File: taqueria-plugin-ligo/common.ts (hunk 1 above)
Comment (mweichert) [reply taqueria-plugin-ligo/common.ts]: OK, I'll adjust. Thanks!

**Review by jchenche — APPROVED** [review]
LGMT!

**Pull request merged**